tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TagMap.java

```diff
@@ -0,0 +1,449 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.metrics.registry;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.function.BiConsumer;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * {@link TagMap} is a {@link SortedMap} implementation optimized for creation performance and memory overhead.
+ * Primarily optimized to retain as little memory as possible, and create small short-lived intermediate objects.
+ *
+ * Note that we expect fairly small tag maps which are iterated over, not used for lookups. Most {@link Map}
+ * methods are implemented, but use a naive linear search rather than a binary search.
+ */
+@SuppressWarnings("JdkObsolete")
+final class TagMap implements SortedMap<String, String> {
+
+    static final TagMap EMPTY = new TagMap(new String[0]);
+
+    private final String[] values;
+    private final int hash;
```

Comment:
I went back and forth on pre-computing the hash since the MetricName implementation already caches the hashCode.

It would be helpful for map wrappers with additional entries, however that case _may_ perform better if we flatten it an create a new TagMap instance with an array two entries larger (assuming relatively few entries, which is generally true). Note that I have not verified this, so take it with a grain of salt.

In the common TaggedMetricSet case which applies additional entries, we never hash the name. It is passed to the metric log subscriber, and only hashed for gauges to ensure we're not blocked recording the same gauge twice.

Reply:
yeah, we could drop the hash field as all the current usages of `TagMap` will now compute the hashCode after creating a `TagMap`, and we now do create a new TagMap when adding an entry in https://github.com/palantir/tritium/pull/1368/files#diff-07d251c64776a9fd9112f154669063beff3ea2c65f4bece8c4bf9c24acce747aR87

Dropped the `hash` field from https://github.com/palantir/tritium/pull/1367

Reply:
Dropped `hash` field here as well